### PR TITLE
[ValueTracking] Improve constant range computation for `shl` and `and`.

### PR DESCRIPTION
**Pull request by goldsteinn**
- [ValueTracking] Add more tests for constant ranges; NFC
- [ValueTracking] Add better support for ConstantRange(Shl)
- [ValueTracking] Add better support for ConstantRange(And)

Proofs: https://alive2.llvm.org/ce/z/rihRSe

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:8602]: I think you need to check the range of C here. Otherwise, the assertions in `APInt::getBitsSetFrom` will fail.
https://github.com/llvm/llvm-project/blob/fa23a2396b6c8d64c3acc47223459ecb75938391/llvm/include/llvm/ADT/APInt.h#L1337-L1355

File: llvm/test/Analysis/ValueTracking/constant-ranges.ll
Comment (dtcxzyw) [inline llvm/test/Analysis/ValueTracking/constant-ranges.ll:101]:
```suggestion
  %x = or i8 %xx, %yy ; thwart complexity-based canonicalization
```

File: llvm/test/Analysis/ValueTracking/constant-ranges.ll
Comment (dtcxzyw) [inline llvm/test/Analysis/ValueTracking/constant-ranges.ll:112]:
```suggestion
  %x = or i8 %xx, %yy ; thwart complexity-based canonicalization
```

File: llvm/test/Analysis/ValueTracking/constant-ranges.ll
Comment (dtcxzyw) [inline llvm/test/Analysis/ValueTracking/constant-ranges.ll:127]:
```suggestion
  %x = or i8 %xx, %yy ; thwart complexity-based canonicalization
```

File: llvm/test/Analysis/ValueTracking/constant-ranges.ll
Comment (dtcxzyw) [reply llvm/test/Analysis/ValueTracking/constant-ranges.ll:101]: We don't need an extra parameter `%yy` to thwart complexity-based canonicalization. I think `%x = mul i8 %xx, %xx` is enough.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Waiting for additional approval from @nikic.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:8601]: ule -> ult. Shift by bit width is also poison.

**Review by nikic — APPROVED** [review]

**Comment by goldsteinn** [comment]
pushed.

**Pull request closed without merging**
